File: problem2.java
// Problem2:https://leetcode.com/problems/maximal-square/)

// Time Complexity : O(n^2)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nothing as such

class Solution {
    public int maxSumAfterPartitioning(int[] A, int K) {
        int[] memo = new int[A.length];
        memo[0] = A[0];
        int initMax = A[0];

        for (int i = 1; i < K; ++i) {
            if (A[i] > initMax) {
                initMax = A[i];
            }
            memo[i] = (i + 1) * initMax;
        }

        for (int i = K; i < A.length; ++i) {
            int cur = 0;
            int kIntervalMax = A[i];
            for (int j = 1; j <= K; ++j) {
                if (A[i - j + 1] > kIntervalMax) {
                    kIntervalMax = A[i - j + 1];
                }
                cur = memo[i - j] + j * kIntervalMax;

                if (cur > memo[i]) {
                    memo[i] = cur;
                }
            }
        }
        return memo[A.length - 1];
    }
}

File: Problem_2.js
// ## Problem2:(https://leetcode.com/problems/partition-array-for-maximum-sum/)

// Time Complexity : O(nk)
// Space Complexity : O(n)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function (arr, k) {

    let n = arr.length;
    let dp = new Array(n);
    dp[0] = arr[0];

    for (let i = 1; i < arr.length; i++) {
        let maxVal = -Infinity;
        let maxSum = -Infinity;
        // Keep a track of the highest number in the k numbers
        for (let j = 1; j <= k; j++) {
            if (i - j + 1 >= 0) {
                maxVal = Math.max(maxVal, arr[i - j + 1]);
            }
            if (i - j >= -1) {
                let prevSum = i - j === -1 ? 0 : dp[i - j];
                // Calculate maximum sum till now and update
                maxSum = Math.max(maxSum, ((maxVal * j) + prevSum));
            }
        }
        dp[i] = maxSum;
    }

    return dp[n - 1];
};